aidb/vector_database/marqo_vector_database.py

```diff
@@ -0,0 +1,215 @@
+
+  def delete_index(self, index_name: str):
+    '''
+    delete an index
+    '''
+    if index_name in self.index_list:
+      self.marqo_client.delete_index(index_name)
+      # self.index_list.remove(index_name)
+      self.index_list = [i.index_name for i in self.marqo_client.get_indexes()['results']]
+      logger.info("Index '%s' deleted.", index_name)
+
+  def _sanitize_index_name(self, index_name: str) -> str:
+    '''
+    index should start with a capital
+    '''
+    return index_name[0].upper() + index_name[1:]
+  
+  def _check_index_validity(self, index_name: str):
+    if index_name not in self.index_list:
+      raise Exception(f'Couldn\'t find index {index_name}, please create it first')
+    return index_name
+  
+  def generate_uuid5(self, x): 
+    return str(uuid.uuid5(uuid.NAMESPACE_DNS, str(x)))
+  
+  def insert_data(self, index_name: str, data: pd.DataFrame):
+    '''
+    insert data into an index
+    '''
+    index_name = self._check_index_validity(index_name)
+
+    data = data.to_dict(orient="records")
+    # Data will have list of dictionaries 
+    # with 'id' and 'values' as keys.
+    
+    mod_data = []
+    for d in data:
+      # print(d)
+      _id = str(d['id'])
+      _vec = d['values']
+      mod_data.append(
+        {
+          "_id": _id,
+          "aidb_data":{
+            "vector": _vec
+          }
+        }
+      )
+    response = self.marqo_client.index(
+        index_name=index_name
+        ).add_documents(
+          documents=mod_data,
+          mappings={
+            "aidb_data":{
+              "type": "custom_vector"
+            }
+          },
+          tensor_fields=["aidb_data"],
+          auto_refresh=True
+        )
+    
+  def get_embeddings_by_id(self, index_name: str, ids: np.ndarray, reload = False) -> np.ndarray:
+    '''
+    Get data by id and return results
+    '''
+    if reload:
+      self.load_index()
+    index_name = self._check_index_validity(index_name)
+    result = []
+    for id in ids:
+      fetch_response = self.marqo_client.index(
+        index_name).get_document(document_id=str(id), expose_facets=True)
+      result.append(fetch_response['_tensor_facets'][0]['_embedding'])
+    return np.array(result)
+  
+  def query_by_embedding(
+      self,
+      index_name: str,
+      query_emb_list: np.ndarray,
+      top_k: int = 5,
+  ) -> (np.ndarray, np.ndarray):
+    '''
+    Query nearest k embeddings, return embeddings and ids
+    '''
+    index_name = self._check_index_validity(index_name)
+    
+    all_topk_reps, all_topk_dists = [], []
+    for query_emb in query_emb_list:
+      response = self.marqo_client.index(index_name).search(
+        q={"dummy": 0},
+        context={
+        'tensor':[{"vector": list(query_emb), "weight" : 1}],
+        }, limit=top_k
+      )
+      ids = []
+      dists = []
+      for res in response['hits']:
+        ids.append(res['_id'])
+        dists.append(1-res['_score']) # Because it is similarity score
```

Comment:
Is this correct for all similarity function or only cosine similarity?

Reply:
Yes I guess, tried it for other metrics, it gives a similarity score between 0 and 1 each time.

Reply:
I saw some scores in documents are larger than 1.

Reply:
Okay, taking a look.